### PR TITLE
[IA-4826] [IA-4785] Add proper notifications and run Azure test on PRs

### PR DESCRIPTION
**Pull request by jdcanas**
This PR:
- Adds code to run azure automation  tests on all PRs
- Fixes slack notifications so we receive a notification in #ia-notification-test for failures on dev promotion and in #dsde-qa for failures from the staging promotion webhook

### CODE REVIEW

File: .github/workflows/azure_automation_test.yml [hunk 1]
```diff
@@ -289,21 +298,34 @@ jobs:
...
       - create-bee-workflow
       - leo-automation-test-job
       - destroy-bee-workflow
-    if: ${{ github.event_name == 'schedule' && failure() }} # Want to notify regardless of which step fails
     steps:
-      - name: Notify slack
-        uses: slackapi/slack-github-action@v1.23.0
+      - id: notify-ia-on-failures
+        name: Notify ia on failures
+        if: ${{ failure() && github.ref_name == 'develop' }} # Want to notify regardless of which step fails.
```
Comment (cpate4) [inline .github/workflows/azure_automation_test.yml:340]: because you have PRs enabled on develop for this - is this what you are wanting? 
from the documentation:
> For pull requests, the format is <pr_number>/merge.

Comment (cpate4) [reply .github/workflows/azure_automation_test.yml:340]: I believe as is, this will only run on committing to develop (and now that I think of it, not really sure what pull-request with a branch of develop would actually do, outside of maybe PRs from develop?) - thinking out loud

**Review by cpate4 — APPROVED** [review]

File: .github/workflows/azure_automation_test.yml (hunk 1 above)
Comment (jdcanas) [reply .github/workflows/azure_automation_test.yml:340]: So there are 2 relevant sections in the `on` block, one is for `pull_request` (i.e. on every commit for PRs), on is `push`. This block should only trigger when the branch it is running against is develop (from the `push` event, not PRs). In other words, every time we merge/commit something to the develop branch itself we get an additional test run and the report to our channel. The idea here is, when the suite is actually stable, anytime it fails against the develop branch its a red flag. 

```
  pull_request:
    branches: [ develop ]
    paths-ignore: [ '**.md' ]
  push:
    branches: [ develop ]
    paths-ignore: [ '**.md' ]
```

**Review by lucymcnatt — APPROVED** [review]

File: .github/workflows/azure_automation_test.yml [hunk 2]
```diff
@@ -1,19 +1,19 @@
-# if you want to debug this and trigger it on PRs against dev
-# uncomment the 2 pull_request lines below
-
 name: Azure automation tests
```
Comment (jyang-broad) [inline .github/workflows/azure_automation_test.yml:3]:
```suggestion
run-name: ${{ inputs.run-name }}

```

File: codecov.yml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 codecov:
-  require_ci_to_pass: yes
+  require_ci_to_pass: true
```
Comment (jdcanas) [inline codecov.yml:2]: this wasn't even a valid option in the config yaml: https://docs.codecov.com/docs/common-recipe-list#see-coverage-on-ci-test-failures

Comment (lucymcnatt) [reply codecov.yml:2]: lol good catch, since when is yes/no valid for a boolean param...

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureRuntimeSpec.scala [hunk 4]
```diff
@@ -172,12 +172,15 @@ class AzureRuntimeSpec
...
+          s"AzureRuntimeSpec: runtime ${workspaceId}/${runtimeName.asString} delete monitor result: $monitorDeleteResult"
         )
         _ = monitorDeleteResult.getStatus() shouldBe ClusterStatus.DELETED
 
         diskAfterRuntimeDelete <- getDisk
-        _ = diskAfterRuntimeDelete.getStatus shouldBe DiskStatus.DELETED
+        _ <- loggerIO.info(
+          s"AzureRuntimeSpec: disk $workspaceId/${diskAfterRuntimeDelete.getName} delete monitor result: $diskAfterRuntimeDelete"
+        )
+        _ = diskAfterRuntimeDelete.getStatus should (be(DiskStatus.DELETED) or be(DiskStatus.DELETING))
```
Comment (lucymcnatt) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureRuntimeSpec.scala:183]: This should be fine, since occasionally disks take longer than VMs to delete on Azure

File: .github/workflows/leo-build-tag-publish-and-run-tests.yml [hunk 5]
```diff
@@ -182,6 +197,32 @@ jobs:
               "custom-version-json": "${{ steps.render-leo-version.outputs.custom-version-json }}"
             }'
 
+  azure-automation-tests:
+    runs-on: ubuntu-latest
+    # TODO: fix the  `ref`
```
Comment (lucymcnatt) [inline .github/workflows/leo-build-tag-publish-and-run-tests.yml]: what needs to be done to fix the `ref`?

File: .github/workflows/leo-build-tag-publish-and-run-tests.yml [hunk 6]
```diff
@@ -9,14 +8,22 @@ on:
...
+    inputs:
+      terra-env:
+        description: 'Terra env this test run is triggered for. This is mainly used to indicate this workflow was triggered from a staging run'
+        required: true
+        default: 'develop'
+        type: string
 
 env:
-  BEE_NAME: '${{ github.event.repository.name }}-${{ github.run_id }}-${{ github.run_attempt}}-dev'
+  BEE_NAME: '${{ github.event.repository.name }}-${{ github.run_id }}-${{ github.run_attempt}}-gcp-dev'
```
Comment (lucymcnatt) [inline .github/workflows/leo-build-tag-publish-and-run-tests.yml:20]: should this be azure for the azure tests?

File: .github/workflows/azure_automation_test.yml [hunk 7]
```diff
@@ -1,19 +1,20 @@
...
+run-name: ${{ inputs.run-name }}
 
 on:
-#  pull_request:
-#    branches: [ develop ]
-  schedule:
```
Comment (lucymcnatt) [inline .github/workflows/azure_automation_test.yml:9]: do we not need to run on a schedule anymore? I'm fine with it, just not sure what the QA requirements are

File: codecov.yml (hunk 3 above)
Comment (jdcanas) [reply codecov.yml:2]: my IDE told me :)

Comment (lucymcnatt) [reply codecov.yml:2]: and TG for IDEs

File: .github/workflows/leo-build-tag-publish-and-run-tests.yml (hunk 5 above)
Comment (jdcanas) [reply .github/workflows/leo-build-tag-publish-and-run-tests.yml]: Ah, I had to hardcode the ref here to be this branch since the changes to the workflow file were on this branch, now it points to develop again correctly. The TODO can go now.

There already exists a way to run azure automation tests against a specific branch with a separate param for the branch the workflow file to run the tests is sourced from, but this change will mean any PR that runs azure automation tests will take the workflow file to run the tests from develop (and the app code and test code will be from the PR branch build).  

tl;dr Very confusing since there are **three** relevant branches, each individually configurable if you run it from workflow-dispatch manually, but it is fixed:
- the branch the workflow file running the test code is from (will always be develop for PRs but is configurable if you manually want to trigger the azure test workflow)
- the branch the automation code running in the GHA node is from (will be the HEAD commit of PR) 
- the branch the leo code deployed to a bee is from (will be the HEAD commit of PR)

File: .github/workflows/azure_automation_test.yml (hunk 7 above)
Comment (jdcanas) [reply .github/workflows/azure_automation_test.yml:9]: Yeah, the scheduling was just to give us regular data on test runs. The QA requirement is running them on the beehive hook from staging, and we will be getting our regular data from the run that occurs whenever a PR is merged to develop

**Comment by lucymcnatt** [comment]
I'm seeing an error in the most recent Azure test run on this PR: `Unexpected inputs provided: ["terra-env", "bee-name", "is-triggered-from-pr-dispatch", "run-name"]`

**Comment by jdcanas** [comment]
> I'm seeing an error in the most recent Azure test run on this PR: `Unexpected inputs provided: ["terra-env", "bee-name", "is-triggered-from-pr-dispatch", "run-name"]`

Yeah, this will fail due to the `ref` comment above, the final code is triggering the workflow file in dev with these inputs (which it does not take until it is merged)

File: codecov.yml (hunk 3 above)
Comment (cpate4) [reply codecov.yml:2]: python has a (read: had a) now deprecated built-in library function (`strtobool`) for doing just that - converting from yes/no (and other string values) to `True` and `False` - 😬

File: .github/workflows/azure_automation_test.yml [hunk 8]
```diff
@@ -158,14 +193,16 @@ jobs:
...
+  # Here we specify we need `create-bee-workflow`, but `if: always()` ensures this is run even if that step is skipped
   leo-automation-test-job:
     runs-on: ubuntu-latest
     needs:
       - init-github-context
-      - create-bee-workflow
       - params-gen
       - attach-billing-project-to-landing-zone-workflow
+      - create-bee-workflow
+    if: always()
```
Comment (lucymcnatt) [inline .github/workflows/azure_automation_test.yml:205]: if we have the `always()` here, is the `needs` section necessary, since it'll run no matter what?

**Review by lucymcnatt — APPROVED** [review]

File: .github/workflows/azure_automation_test.yml [hunk 9]
```diff
@@ -85,10 +102,25 @@ jobs:
...
+          ENV=local
+          BEE_NAME_TEMP=${{ env.BEE_NAME }}
+          IS_TRIGGERED_FROM_PR_DISPATCH=false
+          if ${{ github.event_name == 'workflow_dispatch' }}; then
+            ENV=${{ inputs.terra-env }}
+            IS_TRIGGERED_FROM_PR_DISPATCH=${{ inputs.is-triggered-from-pr-dispatch }}
+            if ${{ inputs.is-triggered-from-pr-dispatch == 'true' }}; then
+              BEE_NAME_TEMP=${{ inputs.bee-name }}
+            fi
+          fi
```
Comment (cpate4) [inline .github/workflows/azure_automation_test.yml:115]: because this is a shell ... 
do we need to wrap this in any way? we are intermixing GitHub expressions with shell scripting

**Review by cpate4 — APPROVED** [review]
